File: solutions/ruby/saddle-points/1/saddle_points.rb
module Grid
  def self.saddle_points(input)
    rows = input
    cols = input.transpose
    saddle_points = []

    rows.each.with_index do |row, row_index|
      row.each.with_index do |row_val, column_index|
        col_val = cols[column_index][row_index]
        if row_val == row.max && col_val == cols[column_index].min
          saddle_points << { "row" => row_index + 1, "column" => column_index + 1 }
        end
      end
    end
    saddle_points
  end
end

File: solutions/python/saddle-points/1/saddle_points.py
def saddle_points(matrix: list[list[int]]) -> list[object]:
    if not all(len(row) == len(matrix[0]) for row in matrix):
        raise ValueError('irregular matrix')

    rows = matrix
    *cols, = map(list, zip(*rows))
    retval = []

    for y, row in enumerate(rows):
        for x, col in enumerate(cols):
            if min(col) == col[y] and max(row) == row[x]:
                retval.append({'column': x+1, 'row': y+1})

    return retval
